[8.x] deps: V8: backport 20 CPU profiler commits from upstream

psmarshall opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

@ofrobots 

[cpu-profiler] Fix bugs and add tests for JITLineInfoTable
https://chromium.googlesource.com/v8/v8/+/4feb5ce7fd5ef8c933f3f5dff2eca1173f85c1e9

[cpu-profiler] Fix incorrect line number calculation.
https://chromium.googlesource.com/v8/v8/+/ddb2856f39632f9e9f623d3cdb4600e636172031

[cpu-profiler] Use std::unordered_map for hashmaps.
https://chromium.googlesource.com/v8/v8/+/35985ce6abc80b85264fe3b87b246fed5f1806e6

[cpu-profiler] Do not store CodeEntries between profiling sessions.
https://chromium.googlesource.com/v8/v8.git/+/8ec48b2117b8092c4956f1ee11a0c85bec3ba1f8

[cpu-profiler] Remove name_prefix field from CodeEntry
https://chromium.googlesource.com/v8/v8.git/+/6f72af25fe43218b60c68129073ddcddb631566e

[cpu-profiler] Extract rare used fields of CodeEntry to an optional object.
https://chromium.googlesource.com/v8/v8.git/+/fcc1ebb55aab38013855834f556f6e874e0eb8b3

[profiler] Refactoring: decouple StringsStorage from Heap object.
https://chromium.googlesource.com/v8/v8/+/a31320f59c911a277566d6c2fa0b0f2ac83e0748

[cpu-profiler] Add a HandleScope to limit memory consumption.
https://chromium.googlesource.com/v8/v8.git/+/3e9f8a4f635e2d946651d6a4df81378266f32dc9

[cpu-profiler] Lazily create CPU profiler.
https://chromium.googlesource.com/v8/v8/+/1426ea1d6d45be0b4d9476bdb5bf3f27cfe578a0

[cpu-profiler] turn several std::map's into unordered_map's.
https://chromium.googlesource.com/v8/v8/+/3ed5dfb8a3cbc7aa0017bd01c2fdd6227485b8ad

[cpu-profiler] Eagerly delete not used CodeEntry'es
https://chromium.googlesource.com/v8/v8.git/+/c6c28f7a412a88df12055e953630a9e93cc64d49

[cpu-profiler] Move bailout reason into rare_info struct
https://chromium.googlesource.com/v8/v8.git/+/29ea4d1ef5360e71c61ecf8db6a5a0a0c3391fd1

[cpu-profiler] Save space in the SourcePositionTable by using a vector.
https://chromium.googlesource.com/v8/v8.git/+/1cb19f0e0a93adbac8c11bc906f951bd8098722d

[cpu-profiler] Only store deopt inline frames for functions that need it
https://chromium.googlesource.com/v8/v8.git/+/0bfcbdd4726920755e51dab28c18ab93e050819b

[cpu-profiler] Add a new profiling mode with a more detailed call tree.
https://chromium.googlesource.com/v8/v8.git/+/ecae80cdb350dde1e654c531b56f5b6c44dc8c77

[cpu-profiler] Reuse free slots in code_entries_
https://chromium.googlesource.com/v8/v8.git/+/3e1126bf15e62c433c4e9cb21316d182f691c63a

[cpu-profiler] Use instruction start as the key for the CodeMap
https://chromium.googlesource.com/v8/v8.git/+/ba752ea4c50713dff1e94f45a79db3ba968a8d66

ofrobots commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/15653/
Will launch V8-CI once #21433 is resolved.

ofrobots commented:
The V8 version in `v8-version.h` will need to be bumped.

/cc @nodejs/v8-update

ofrobots commented:
Relaunched suspected flakes/infra issues:
* linuxone build: https://ci.nodejs.org/job/node-test-commit-linuxone/2616/
* linter: https://ci.nodejs.org/job/node-test-linter/20332/

psmarshall commented:
Thanks, bumped v8-version

ofrobots commented:
CI is green. V8-CI is functional now, so here's a launch: https://ci.nodejs.org/view/All/job/node-test-commit-v8-linux/1504/

/fyi @nodejs/diagnostics: This improves the memory overhead of the CPU profiler. These fixes are available on 10 (all commits except the last one are in V8 6.7), and this PR backports them to Node 8.

/cc @nodejs/lts.

hashseed commented:
> The V8 version in `v8-version.h` will need to be bumped.

I thought we bump the embedder string for backmerges.

MylesBorins commented:
landed in b57ad02d74

ofrobots commented:
Oops, I messed. I intended to add this comment here, but put it on the wrong issue: https://github.com/nodejs/node/pull/21334#issuecomment-402254851.

> Actually, I was mistaken, these fixes not in 10 just yet. They are in V8 6.8 (rather than 6.7). Let's hold off on landing this until a 10.x release picks them up.

ofrobots commented:
@MylesBorins is it possible to revert these commits from 8.x?

MylesBorins commented:
I've backed the commit out of staging and the 8.12.0 proposal

ofrobots commented:
Thanks. I've added the `in progress` label to make sure it doesn't get re-landed until it is time.

psmarshall commented:
Note to self: Add this one too if it's fine on 10, it's the new API:
https://chromium.googlesource.com/v8/v8/+/ecae80cdb350dde1e654c531b56f5b6c44dc8c77

hashseed commented:
I don't have any technical objections against the new API.

psmarshall commented:
I added 3 more commits - one full of bugfixes, one with the new API, and one that was necessary for easier merging.

MylesBorins commented:
@psmarshall would you be able to rebase? Are these commits now out in the wild as of the latest 8.x?

psmarshall commented:
They aren't released yet: still waiting on https://github.com/nodejs/node/pull/22028. I'll rebase :+1:

Flarna commented:
@psmarshall any chance that this is continued?

I created #23070 yesterday and according to @ofrobots this PR should fix this.

psmarshall commented:
@Flarna I've just rebased this and the changes have landed in Node 10 so we are good to go with this PR :+1: 

@MylesBorins I'm ready to land :rocket:

MylesBorins commented:
Let's do one more round of CI and then land

V8-CI: https://ci.nodejs.org/job/node-test-commit-v8-linux/1714/

BethGriggs commented:
Another V8-CI: https://ci.nodejs.org/job/node-test-commit-v8-linux/1724

MylesBorins commented:
looks like compilation errors

```
11:22:43 make[2]: *** [/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ubuntu1604-intel-64/v8test/v8test/deps/v8/out/x64.release/obj.target/cctest/test/cctest/test-log.o] Error 1
11:22:43 make[2]: *** Waiting for unfinished jobs....
11:22:43 12 errors generated.
11:22:43 test/cctest/cctest.target.x64.release.mk:370: recipe for target '/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ubuntu1604-intel-64/v8test/v8test/deps/v8/out/x64.release/obj.target/cctest/test/cctest/test-cpu-profiler.o' failed
11:22:43 make[2]: *** [/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ubuntu1604-intel-64/v8test/v8test/deps/v8/out/x64.release/obj.target/cctest/test/cctest/test-cpu-profiler.o] Error 1
11:22:43 2 errors generated.
11:22:43 test/cctest/cctest.target.x64.release.mk:370: recipe for target '/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ubuntu1604-intel-64/v8test/v8test/deps/v8/out/x64.release/obj.target/cctest/test/cctest/test-profile-generator.o' failed
11:22:43 make[2]: *** [/home/iojs/build/workspace/node-test-commit-v8-linux/nodes/ubuntu1604-intel-64/v8test/v8test/deps/v8/out/x64.release/obj.target/cctest/test/cctest/test-profile-generator.o] Error 1
11:22:56 rm bdc92efe44eca1954fdff70fd086c7e138d0fbb1.intermediate
11:22:56 Makefile:319: recipe for target 'x64.release' failed
11:22:56 make[1]: *** [x64.release] Error 2
11:22:56 Makefile:215: recipe for target 'v8' failed
11:22:56 make: *** [v8] Error 2
11:22:56 Build step 'Execute shell' marked build as failure
11:22:56 Recording test results
```

/cc @psmarshall @hashseed

psmarshall commented:
@MylesBorins Thanks, the cctest build was failing, I fixed a few issues there and it passes locally now. I added this backport (https://github.com/nodejs/node/pull/22688) that we did to v10.x, too.

rvagg commented:
v8 tests https://ci.nodejs.org/job/node-test-commit-v8-linux/1729/
full ci https://ci.nodejs.org/job/node-test-pull-request/17629/

rvagg commented:
debian8 failure unrelated (although concerning), this seems good to go

Trott commented:
Linux CI rebuild: https://ci.nodejs.org/job/node-test-commit-linux/22248/

I think (once that is green/yellow) this is good to land? Or...?   /ping @BethGriggs @MylesBorins

BethGriggs commented:
Landed in 6ae7b3daad v8.x-staging
